sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -453,7 +453,8 @@ abstract class CastBase extends UnaryExpression with TimeZoneAwareExpression wit
   // TimestampConverter
   private[this] def castToTimestamp(from: DataType): Any => Any = from match {
     case StringType =>
-      buildCast[UTF8String](_, utfs => DateTimeUtils.stringToTimestamp(utfs, zoneId).orNull)
+      buildCast[UTF8String](_,
+        utfs => DateTimeUtils.stringToTimestampAnsi(utfs, zoneId, ansiEnabled).orNull)
```

Comment:
We should call `stringToTimestampAnsi` when ansi mode is on, and call the original method when ansi mode is off.